### ISSUE TITLE
problem importing numpy extension multiarray

### ISSUE THREAD

**Issue by prenaut41** [ContinuumIO/anaconda-issues#10561]
From what I've seen on numpy issues, it sounds like the source of this problem is more likely with the Anaconda distribution.   
The error below happens when I try to run a script importing numpy.  Also happens if I simply try to import numpy in python terminal.  What is the fix for this?  

-------------------------------------
**environment:**
- OS: Windows 7, 64-bit
- Anaconda3-5.3.1-Windows-x86_64
- Python 3.7.0
- numpy-1.15.1

-------------------------------------
**Error:**
Traceback (most recent call last):
  File "C:\apps\Anaconda3\lib\site-packages\numpy\core\__init__.py", line 16, in
 <module>
    from . import multiarray
ImportError: DLL load failed: The specified module could not be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "\\nvaxfile\rockville_shared\Tech Transfer\qcat\testmethods\RSV SDS-PAGE\
RSVSDSPAGEDataScraper_20190124.py", line 7, in <module>
    import numpy as np
  File "C:\apps\Anaconda3\lib\site-packages\numpy\__init__.py", line 142, in <mo
dule>
    from . import add_newdocs
  File "C:\apps\Anaconda3\lib\site-packages\numpy\add_newdocs.py", line 13, in <
module>
    from numpy.lib import add_newdoc
  File "C:\apps\Anaconda3\lib\site-packages\numpy\lib\__init__.py", line 8, in <
module>
    from .type_check import *
  File "C:\apps\Anaconda3\lib\site-packages\numpy\lib\type_check.py", line 11, i
n <module>
    import numpy.core.numeric as _nx
  File "C:\apps\Anaconda3\lib\site-packages\numpy\core\__init__.py", line 26, in
 <module>
    raise ImportError(msg)
ImportError:
Importing the multiarray numpy extension module failed.  Most
likely you are trying to import a failed build of numpy.
If you're working with a numpy git repo, try `git clean -xdf` (removes all
files not under version control).  Otherwise reinstall numpy.

Original error was: DLL load failed: The specified module could not be found.

------------------------------------------
I don't have any other python environments installed on this machine - just the one from Anaconda.  
Interestingly, I can import numpy just fine if I'm in Jupyter Notebook, but not from the python terminal and not in VS Code.

**Comment by prenaut41** [comment]
As additional information, I have no such problem on my development environment:
- OS: Windows 7, 64-bit
- Anaconda3-4.4.0-Windows-x86_64
- Python 3.6.1
- numpy-1.12.1

**Comment by csoja** [comment]
This looks to be a duplicate of https://github.com/spyder-ide/spyder/issues/7357
Please upgrade the failing environment to Anaconda 2018.12.

**Comment by msarahan** [comment]
Upgrading alone probably won't fix the issue.  Please read the issue that @csoja posted.  It describes the root cause: some other software vendor has been naughty and put DLLs in a global location.  Those DLLs are preventing Anaconda from working properly.  We're working on workarounds, but fundamentally the problem is other people abusing the way that Windows works, and breaking anyone else who uses the same libraries (but different versions) than them.

**Comment by prenaut41** [comment]
OK.  Any suggestions on what I should try in the meantime?  

- should I bother upgrading to Anaconda 2018.12?  If so, what is the best way to do that cleanly?   
- conda install mkl=2018.0.2
- Is this a possible solution?: [https://stackoverflow.com/a/50241711/3341533](url)

Thanks

**Comment by mingwandroid** [comment]
Hi @prenaut41,

Can you try:

`conda update python -c rdonnelly`

**Comment by prenaut41** [comment]
@mingwandroid - I don't really feel very comfortable about installing anything other than an official distribution or package in this environment that is experiencing the problem.

**Comment by msarahan** [comment]
@mingwandroid is an Anaconda, Inc. employee.  What he's asking you to test is a preview release of Python which has been patched to find our DLLs earlier than ones in System32.  It might fix your problem.  If you can't try this, you'll just need to wait, or go read the other issues.  We can't offer you any further free support.  The information on conflicting DLLs is out there, and you need to read it and understand how your system is manifesting the issue.

**Comment by mingwandroid** [comment]
Further to what @msarahan said, you can also create new environments if you are worried about breaking any existing environments (though they sound pretty broken to me for you already).

**Comment by prenaut41** [comment]
Yeah, I get that.  But I can't take that kind of risk in this particular environment and I would be breaking the trust with our IT dept .  I would be exposing too much.  It's not about the anaconda/python environment, but the exposure of our entire network.  I could see if our IT dept could stand up a "sandbox" environment which is more isolated and first test it out there first before then trying it on the machine that is actually experiencing the problem, but I'm not sure if I'll be able to do that.

**Comment by prenaut41** [comment]
Confirmed that the problem still exists after updating anaconda with anaconda version "2018.12" and numpy version "1.15.4".

**Comment by mingwandroid** [comment]
@prenaut41, I told you how to aid in testing the release candidate packages that fix. I'm not sure why you think anything else might fix this; nothing else will short of removing the conflicting DLLs from C:\Windows\System32 (which is more likely to violate an IT policy). I do not understand the risk you speak of or how installing a package from my channel could break trust considering that I have full access to the Anaconda `defaults` channel the same as I do to the `rdonnelly` channel (in fact I have exclusive write access to that channel). To be clear, the package I asked you to test is a release candidate which means if we get enough confirmation that they fix this particular problem (which is the only change made to them), then I will upload these exact binaries to the `defaults` channel.

If you can help me to understand the risk and trust issues you talk about then I'd appreciate it.

**Comment by prenaut41** [comment]
@mingwandroid - I'm working on getting something set up to test out your experimental build.  
I can't explain the risk and trust issues that I have because I don't really understand it myself - it's that lack of understanding that makes me want to be careful installing something from a special channel based on a github chat.

**Comment by prenaut41** [comment]
Regarding the conflicting DLLs, I don't think it's the MKL library because I don't find any mkl-related DLLs outside of the Anaconda folders when I search the machine.  The error message doesn't tell me what specific DLL is a problem, so I'm still trying to get a clue on this aspect.

**Comment by mingwandroid** [comment]
It doesn't have to be MKL, it can be any DLL that conda/Python depends upon.

**Comment by mingwandroid** [comment]
If you have process hacker installed you could probably use that to see which DLLs get loaded.

**Comment by prenaut41** [comment]
@mingwandroid - I was able to install and test out the custom build, but unfortunately I still get exactly the same error message.  

 ---------------------------------------------------------------------------------------
**Upgrade to custom build:**

(base) C:\Users\Administrator.INFORUTILTEST>conda update python -c rdonnelly
Collecting package metadata: done
Solving environment: done

## Package Plan ##

  environment location: C:\apps\Anaconda3

  added / updated specs:
    - python

The following packages will be downloaded:
    package                    |            build
    ---------------------------|-----------------
    anaconda-custom            |           py37_0           6 KB
    python-3.7.2               |       h8c8aaf0_1        17.8 MB  rdonnelly
    qt-5.9.7                   |   vc14h73c81de_0        92.3 MB  rdonnelly
    ------------------------------------------------------------
                                           Total:       110.1 MB

The following packages will be UPDATED:
  ca-certificates                              2018.03.07-0 --> 2018.12.5-0
  python                 pkgs/main::python-3.7.1-h8c8aaf0_6 --> rdonnelly::python-3.7.2-h8c8aaf0_1

The following packages will be SUPERSEDED by a higher-priority channel:
  qt                                              pkgs/main --> rdonnelly

The following packages will be DOWNGRADED:
  anaconda                                   2018.12-py37_0 --> custom-py37_0
Proceed ([y]/n)? y

Downloading and Extracting Packages
anaconda-custom      | 6 KB      | #################################### | 100%
qt-5.9.7             | 92.3 MB   | #################################### | 100%
python-3.7.2         | 17.8 MB   | #################################### | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done



-----------------------------------------------------------------------------------------------
**Repeat of Error message on attempting to import numpy:**

C:\>python
Python 3.7.2 (default, Jan 27 2019, 19:17:17) [MSC v.1915 64 bit (AMD64)] :: Ana
conda custom (64-bit) on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
Traceback (most recent call last):
  File "C:\apps\Anaconda3\lib\site-packages\numpy\core\__init__.py", line 16, in
 <module>
    from . import multiarray
ImportError: DLL load failed: The specified module could not be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\apps\Anaconda3\lib\site-packages\numpy\__init__.py", line 142, in <mo
dule>
    from . import add_newdocs
  File "C:\apps\Anaconda3\lib\site-packages\numpy\add_newdocs.py", line 13, in <
module>
    from numpy.lib import add_newdoc
  File "C:\apps\Anaconda3\lib\site-packages\numpy\lib\__init__.py", line 8, in <
module>
    from .type_check import *
  File "C:\apps\Anaconda3\lib\site-packages\numpy\lib\type_check.py", line 11, i
n <module>
    import numpy.core.numeric as _nx
  File "C:\apps\Anaconda3\lib\site-packages\numpy\core\__init__.py", line 26, in
 <module>
    raise ImportError(msg)
ImportError:
Importing the multiarray numpy extension module failed.  Most
likely you are trying to import a failed build of numpy.
If you're working with a numpy git repo, try `git clean -xdf` (removes all
files not under version control).  Otherwise reinstall numpy.

Original error was: DLL load failed: The specified module could not be found.

**Comment by prenaut41** [comment]
I'm trying to dig further into what the missing DLL is.  Using Process Monitor, it looks like mkl_rt.dll cannot be found:

![image](https://user-images.githubusercontent.com/16271316/52026682-bb91b300-24d6-11e9-9d4c-1e1e1019649d.png)

The interesting thing is that this file does actually exist in  I don't understand this though, since this file actually does exist in "Anaconda3\Library\bin" along with a bunch of other mkl files, but this path is not on the search path for some reason.  

![image](https://user-images.githubusercontent.com/16271316/52026841-5094ac00-24d7-11e9-9c8b-53a9617a3cb7.png)

After copying mkl_rt.dll into numpy\core, then I get a slightly different error message with a fatal crash of python, saying that mkl_intel_thread.dll couldn't be loaded.  After copying that dll over as well, then I get a windows msg indicating a system error because libiomp5md.dll is missing, with the python terminal stating it cannot load mkl_intel_thread.dll, which I had just added to the path:  

![image](https://user-images.githubusercontent.com/16271316/52027303-e7159d00-24d8-11e9-8727-408339e15ad2.png)

![image](https://user-images.githubusercontent.com/16271316/52027567-dd406980-24d9-11e9-9dd6-98dc423e9fb9.png)

So I copied libiomp5md.dll from Anaconda3\Library\bin to the parent Anaconda3\ folder since there were no other Anaconda-related folders in the search path for this one.  
Then it failed looking for mkl_core.  

So I went back to this post from stack overflow and tried copying all the listed dlls to \numpy\core.  I tried this earlier, but I did not realize that I also had to make libiomp5md.dll available, which was not on this list:
https://stackoverflow.com/a/50241711/3341533

This finally worked.  Now I can import and use numpy.  Finally.  

But what is really going on here?  I'm new to trying to diagnose this DLL Hell, but this doesn't seem to me to be the result of a 3rd party software installing conflicting versions of DLLs in global folders.  The only copies I have of these DLLs are from the Anaconda package.  It seems like all the DLLs are there, but for some reason they are not getting found because the locations for these are not being included in the search paths.  Why is that?  
Is something wrong with my Anaconda installation?

**Comment by mingwandroid** [comment]
Hi @prenaut41, this is very strange. Just to be clear, are you using the `Anaconda Prompt` here? You haven't done something like manually adding `Anaconda` to your `PATH` have you?

If you have the new build of Python installed (without the workaround of having copied the DLLs into `Anaconda3`) from the `Anaconda Prompt`, can you enter:

```
set CONDA_DLL_SEARCH_MODIFICATION_DEBUG=1
python.exe -c "import numpy"
```

.. and paste the output here please?

**Comment by prenaut41** [comment]
@mingwandroid - 
I normally don't mess around with PATH variable unless I really need to.  I see that Anaconda3 is in PATH, though.  Is that not done by the installer?  Maybe I did add it so that the python.exe would be found for any ".py" files as the python used by VS Code - this wasn't that long ago, but I can't remember.  

For all the conda commands (updates and such), I was opening a terminal through Anaconda Navigator -> Environments -> base (root) -> Open Terminal.  

When I was trying to run something in python (i.e., testing import numpy), I was opening a regular old windows command prompt and first executing "python" to open a python terminal and then executing "import numpy as np".

-----------------------------------------------------------------------------------
**Here's the output you asked for from the Anaconda Prompt:**

(base) C:\Users\Administrator.INFORUTILTEST>set CONDA_DLL_SEARCH_MODIFICATION_DEBUG=1
(base) C:\Users\Administrator.INFORUTILTEST>python.exe -c "import numpy"

CondaEcosystemModifyDllSearchPath() :: PATH=C:\apps\Anaconda3;C:\apps\Anaconda3\
Library\mingw-w64\bin;C:\apps\Anaconda3\Library\usr\bin;C:\apps\Anaconda3\Librar
y\bin;C:\apps\Anaconda3\Scripts;C:\apps\Anaconda3\bin;C:\apps\Anaconda3\condabin
;C:\apps\Anaconda3;C:\apps\Anaconda3\Library\mingw-w64\bin;C:\apps\Anaconda3\Lib
rary\usr\bin;C:\apps\Anaconda3\Library\bin;C:\apps\Anaconda3\Scripts;C:\Program
Files (x86)\Intel\iCLS Client\;C:\Program Files\Intel\iCLS Client\;C:\Windows\sy
stem32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell
\v1.0\;C:\Program Files\Intel\Intel(R) Management Engine Components\DAL;C:\Progr
am Files (x86)\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\
Intel\Intel(R) Management Engine Components\IPT;C:\Program Files (x86)\Intel\Int
el(R) Management Engine Components\IPT;C:\apps\Anaconda3;C:\apps\Microsoft VS Co
de\bin
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Microsoft VS Code
\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Program Files (x86)\In
tel\Intel(R) Management Engine Components\IPT)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Program Files\Intel\In
tel(R) Management Engine Components\IPT)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Program Files (x86)\In
tel\Intel(R) Management Engine Components\DAL)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Program Files\Intel\In
tel(R) Management Engine Components\DAL)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Windows\System32\Windo
wsPowerShell\v1.0)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Windows\System32\Wbem)

CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Windows)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Windows\system32)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Program Files\Intel\iC
LS Client)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\Program Files (x86)\In
tel\iCLS Client)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Scripts
)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Library
\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Library
\usr\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Library
\mingw-w64\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\condabi
n)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Scripts
)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Library
\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Library
\usr\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3\Library
\mingw-w64\bin)
CondaEcosystemModifyDllSearchPath() :: AddDllDirectory(C:\apps\Anaconda3)

(base) C:\Users\Administrator.INFORUTILTEST>

**Comment by mingwandroid** [comment]
> For all the conda commands (updates and such), I was opening a terminal through Anaconda Navigator -> Environments -> base (root) -> Open Terminal.

This should work. 

> When I was trying to run something in python (i.e., testing import numpy), I was opening a regular old windows command prompt and first executing "python" to open a python terminal and then executing "import numpy as np".

This will not work and is not supported.  Use Open Terminal as above or use the Anaconda Prompt.

**Comment by prenaut41** [comment]
Ah.  So that must be my problem.  

I am using this machine to schedule tasks that periodically run various python scripts for some automation.  I set the Windows task manger to run the scripts directly, and in this way I am currently relying on the OS to know to run any .py file with the python interpreter that I had installed along with Ananconda.  I have things working now with that "hacky" copying of dlls to folders where they will be found.

Is there a better way to do this where the interpreter is launched within the context of Anaconda so that everything loads properly before executing the script?
